cmd/juju/cloud/add_test.go

```diff
@@ -187,240 +271,225 @@ func (s *addSuite) TestInteractive(c *gc.C) {
-    regions:
-      london:
-        endpoint: http://london/1.0
-  man:
-    type: manual
-    endpoint: 192.168.1.6
-`[1:])
+	c.Check(numCallsToWrite(), gc.Equals, 1)
 }
 
 func (s *addSuite) TestInteractiveVSphere(c *gc.C) {
-	s.createTestCloudData(c)
-	command := cloud.NewAddCloudCommand()
+	fake := newFakeCloudQueryer()
+	fake.Call("PublicCloudMetadata", []string(nil)).Returns(map[string]cloudfile.Cloud{}, false, nil)
+	fake.Call("PersonalCloudMetadata").Returns(map[string]cloudfile.Cloud{}, nil)
+	vsphereCloud := cloudfile.Cloud{
+		Type:      "vsphere",
+		AuthTypes: []cloudfile.AuthType{"userpass", "access-key"},
+		Endpoint:  "192.168.1.6",
+		Regions: []cloudfile.Region{
+			{
+				Name:     "foo",
+				Endpoint: "192.168.1.6",
+			},
+			{
+				Name:     "bar",
+				Endpoint: "192.168.1.6",
+			},
+		},
+	}
+	fake.Call("ParseOneCloud", []byte("auth-types:\n- userpass\nendpoint: 192.168.1.6\nregions:\n  bar: {}\n  foo: {}\n")).Returns(vsphereCloud, nil)
+	vsphereMetadata := map[string]cloudfile.Cloud{"mvs": vsphereCloud}
+	numCallsToWrite := fake.Call("WritePersonalCloudMetadata", vsphereMetadata).Returns(nil)
+
+	command := cloud.NewAddCloudCommand(fake)
 	err := testing.InitCommand(command, nil)
 	c.Assert(err, jc.ErrorIsNil)
 
 	ctx := &cmd.Context{
-		Dir:    c.MkDir(),
 		Stdout: ioutil.Discard,
 		Stderr: ioutil.Discard,
-		Stdin: strings.NewReader(`
-vsphere
-mvs
-192.168.1.6
-foo
-y
-bar
-n
-`[1:])}
+		Stdin: strings.NewReader("" +
+			/* Select cloud type: */ "vsphere\n" +
+			/* Enter a name for the cloud: */ "mvs\n" +
+			/* Enter the controller's hostname or IP address: */ "192.168.1.6\n" +
+			/* Enter region name: */ "foo\n" +
+			/* Enter another region? (Y/n): */ "y\n" +
+			/* Enter region name: */ "bar\n" +
+			/* Enter another region? (Y/n): */ "n\n",
+		),
+	}
 
 	err = command.Run(ctx)
 	c.Check(err, jc.ErrorIsNil)
 
-	data, err := ioutil.ReadFile(osenv.JujuXDGDataHomePath("clouds.yaml"))
-
-	// Yes, the conversion copies the endpoint to each region, for some reason.
-	c.Assert(string(data), gc.Equals, `
-clouds:
-  homestack:
-    type: openstack
-    auth-types: [userpass, access-key]
-    endpoint: http://homestack
-    regions:
-      london:
-        endpoint: http://london/1.0
-  mvs:
-    type: vsphere
-    auth-types: [userpass]
-    endpoint: 192.168.1.6
-    regions:
-      bar:
-        endpoint: 192.168.1.6
-      foo:
-        endpoint: 192.168.1.6
-`[1:])
+	c.Check(numCallsToWrite(), gc.Equals, 1)
 }
 
 func (s *addSuite) TestInteractiveExistingNameOverride(c *gc.C) {
-	s.createTestCloudData(c)
-	command := cloud.NewAddCloudCommand()
+	fake := newFakeCloudQueryer()
+	fake.Call("PublicCloudMetadata", []string(nil)).Returns(map[string]cloudfile.Cloud{}, false, nil)
+	fake.Call("PersonalCloudMetadata").Returns(homestackMetadata(), nil)
+	manMetadata := map[string]cloudfile.Cloud{"homestack": manualCloud}
+	fake.Call("ParseOneCloud", []byte("endpoint: 192.168.1.6\n")).Returns(manualCloud, nil)
```

Comment:
Right now, if the expected input of Parse changes, you need to change two things in this package: the actual code that produces the yaml, AND this test.

If you didn't mock out Parse here, then when Parse changed, you'd only have to change the actual code that produces the yaml.  This test would still test the interaction between the two, without having to encode knowledge in the *test* about what Parse expects.